### PR TITLE
Fix syncing enum & set types

### PR DESCRIPTION
**Pull request by notfelineit**
This PR implements initial mapping SET and ENUM mysql values to their string values. 

## Context on `Read`
For others not familiar with this connector code, on `Read`, our connector roughly does:
1. Fetch the database schema using our existing `Schema` handler
2. Create a mapping of `column` to serializer; a serializer takes in a mysql value and outputs the value converted to a FiveTran data type
3. On each row,  serialize the mysql value of each column in the row by calling the column's serializer on the underlying mysql value

## Before
Before this change, during the replication phase of `Update`, the connector would send the underlying indices instead of the expected ENUM or SET value(s). Meaning, for a table that looks like:
```
CREATE TABLE `shirts` ( 
  `name` varchar(40) DEFAULT NULL,
  `size` enum('x-small','small','medium','large','x-large') DEFAULT NULL,
  `weekday` set('monday','tuesday','wednesday','thursday','friday','saturday','sunday') DEFAULT NULL )
ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

if a user inserted a new record like:
```
insert into shirts(name, size, weekday)
values ("new-shirt", "medium", "wednesday,friday");
```
the row would be serialized as: `"new-shirt", "3", "3"`

## After
Now we maintain a mapping of indices to values for `ENUM` and `SET` columns during the initial `Schema` read (step 1 in "Context on `Read`"). We use these mappings to produce `ENUM` and `SET` serializers for columns that are of `SET` and `ENUM` type.

If the underlying MySQL value is an integer, the serializer maps the integer to a `SET` or `ENUM` value. Otherwise, it just returns the underlying MySQL value as a string (for `ENUM`) or JSON (for `SET`).

## Follow-up work
After this initial PR is merged, in the mapping logic, it would be better to use a conditional like "if in REPLICATING phase, map, else, use default value"

## Alternatives
We could have parsed and stored the enum/set mapping values from `qr.Fields` in both `onUpdate` and `onRow`: https://github.com/planetscale/fivetran-source/blob/05842c9db98fad6f9498e1a1303327d2f6104aed/lib/connect_client.go#L232 instead of doing so during the [initial schema read](https://github.com/planetscale/fivetran-source/pull/38/files#diff-5f6cd2f8becbc8321cf5dfc6f5a5d369951730ecfb19811a543ba89051722d92R177). Doing so within the initial schema read seemed the least intrusive option currently.

### CODE REVIEW

File: cmd/internal/server/handlers/schema.go [hunk 1]
```diff
@@ -15,5 +15,5 @@ func (Schema) Handle(ctx context.Context, psc *lib.PlanetScaleSource, db *lib.My
 	if err := (*db).BuildSchema(ctx, *psc, schemaBuilder); err != nil {
 		return nil, err
 	}
-	return schemaBuilder.(*fivetranSchemaBuilder).BuildResponse()
+	return schemaBuilder.(*FiveTranSchemaBuilder).BuildResponse()
```
Comment (notfelineit) [inline cmd/internal/server/handlers/schema.go:18]: Had to make `FiveTranSchemaBuilder` public, because in server.go we refer to it when building our schema with metadata!

File: cmd/internal/server/handlers/fivetran_value_converters.go [hunk 2]
```diff
@@ -214,3 +216,51 @@ func GetConverter(dataType fivetransdk.DataType) (ConverterFunc, error) {
 	}
 	return converter, nil
 }
+
+func GetEnumConverter(enumValues []string) (ConverterFunc, error) {
+	return func(value sqltypes.Value) (*fivetransdk.ValueType, error) {
+		parsedValue := value.ToString()
+		index, err := strconv.ParseInt(parsedValue, 10, 64)
+		if err != nil {
```
Comment (notfelineit) [inline cmd/internal/server/handlers/fivetran_value_converters.go:224]: In another PR, it would be good to replace this with a definitive "if not in copying phase, map to value" conditional instead. But this is a step in the right direction I think!

Comment (notfelineit) [reply cmd/internal/server/handlers/fivetran_value_converters.go:224]: Doing so will allow us to successfully parse cases where the user has integer ENUM values `enum("1,",2","3")`. This is not recommended by MySQL but exists. See ["Handling of Enumeration Literals"
](https://dev.mysql.com/doc/refman/8.0/en/enum.html)

File: cmd/internal/server/handlers/fivetran_value_converters.go [hunk 3]
```diff
@@ -214,3 +216,72 @@ func GetConverter(dataType fivetransdk.DataType) (ConverterFunc, error) {
 	}
 	return converter, nil
 }
+
+func GetEnumConverter(enumValues []string) (ConverterFunc, error) {
+	return func(value sqltypes.Value) (*fivetransdk.ValueType, error) {
+		parsedValue := value.ToString()
+		index, err := strconv.ParseInt(parsedValue, 10, 64)
```
Comment (Phanatic) [inline cmd/internal/server/handlers/fivetran_value_converters.go:223]: We should handle a string value of "NULL" here. as it stands right now, a value of `NULL` will fail the serialization, instead we should send a value type of null or the default value of the enum column as we do here : 
https://github.com/planetscale/fivetran-source/blob/05842c9db98fad6f9498e1a1303327d2f6104aed/cmd/internal/server/handlers/schema_aware_serializer.go#L112-L114

> If an ENUM column is declared to permit NULL, the NULL value is a valid value for the column, and the default value is NULL. If an ENUM column is declared NOT NULL, its default value is the first element of the list of permitted values.

**Review by Phanatic — COMMENTED** [review]
Looks great! once we handle NULL values, this should be good to merge.

File: cmd/internal/server/handlers/fivetran_value_converters.go (hunk 3 above)
Comment (notfelineit) [reply cmd/internal/server/handlers/fivetran_value_converters.go:223]: It seems like that conditional `if val.IsNull()` makes it so that the value is set to `fiventrasdk.ValueType_Null`, and the else case [where the enum serializer/set serializer would run](https://github.com/planetscale/fivetran-source/blob/05842c9db98fad6f9498e1a1303327d2f6104aed/cmd/internal/server/handlers/schema_aware_serializer.go#L116), never happens 🤔 is that not the case? 

Locally, it appears that null values are serialized correctly during the replication phase:
Screenshot 2023-12-12 at 8 59 12 AM

Comment (notfelineit) [reply cmd/internal/server/handlers/fivetran_value_converters.go:223]: Unless, we're expecting a string that looks like "NULL"?

Comment (Phanatic) [reply cmd/internal/server/handlers/fivetran_value_converters.go:223]: I think `val.IsNull()` just short circuits this check and  might be all we need for ENUM types. 
The other thing is that this statement isn't true for PlanetScale databases 
> If an ENUM column is declared NOT NULL, its default value is the first element of the list of permitted values.

``` sql
show variables like "%sql_mode%";
+---------------+-----------------------------------------------------------------------------------------------------------------------+
| Variable_name | Value                                                                                                                 |
+---------------+-----------------------------------------------------------------------------------------------------------------------+
| sql_mode      | ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION |
+---------------+-----------------------------------------------------------------------------------------------------------------------+
```

We set [STRICT_TRANS_TABLES](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_strict_trans_tables) to true which will prevent a row from being created with an invalid ENUM value. 
TLDR: what you have is correct, we don't need a special override for NULL enum values.
